include/power/power.h

```diff
@@ -64,6 +64,43 @@ extern unsigned char sys_pm_idle_exit_notify;
  * @{
  */
 
+/**
+ * Power managment notifier struct
+ *
+ * This struct contains callbacks that are called when the target enters and
+ * exits power states.
+ */
+struct pm_notifier {
+	/**
+	 * Application defined function for doing any target specific operations
+	 * for power state entry.
+	 */
+	void (*state_entry)(enum power_states state);
+	/**
+	 * Application defined function for doing any target specific operations
+	 * for power state exit.
+	 */
+	void (*state_exit)(enum power_states state);
+};
+
+/**
+ * @brief Statically define a power management notifier.
+ *
+ * Define a power managment notifier.
+ *
+ * @param name Name of the pm_notifier variable.
+ * @param entry_fn Function to be called when platform enters a power state
+ * @param exit_fn  Function to be called when platform exits a power state
+ */
+#define PM_NOTIFIER_DEFINE(name, entry_fn, exit_fn)		\
+	static Z_STRUCT_SECTION_ITERABLE(pm_notifier, name) =	\
```

Comment:
We should get some clarity on when objects are aggregated into an array by the linker, and when they are not.

If `pm_notifier` were a kernel object the aggregation would make sense.  Otherwise this API doesn't seem ideal, as it assumes that all notifiers are identified at build time and are enabled unconditionally at boot.  This forces dependencies: a module may need to set things up before its notifier is ready to be called, which won't have been done if the system does any suspension before the module is initialized (which may not happen at boot).

I think a register/unregister API that adds/removes a notifier is more flexible and has only a small cost (a link node in the notifier structure).

Reply:
They can be initialized with NULL pointer callbacks in this case, we can go with the linked list approach, I have nothing against that. What I want here is confirm whether or not makes sense allow multiple notifiers.
Btw, make it kernel objects may make sense if applications want to use it from user mode and I have the feeling that this will be a valid case.